File: DatabaseHelper.vb
Imports System.Data.SQLite
Imports System.Data

Public Class DatabaseHelper
    Implements IDisposable

    Private _connection As SQLiteConnection
    Private _connectionString As String
    Private _isDisposed As Boolean = False

    Public Sub New(connectionString As String)
        _connectionString = connectionString
    End Sub

    Public ReadOnly Property Connection As SQLiteConnection
        Get
            If _connection Is Nothing Then
                _connection = New SQLiteConnection(_connectionString)
            End If
            If _connection.State <> ConnectionState.Open Then
                _connection.Open()
            End If
            Return _connection
        End Get
    End Property

    Public Sub EnsureConnection()
        Dim unused = Connection
    End Sub

    ''' <summary>
    ''' 执行非查询 SQL，使用显式类型的参数
    ''' </summary>
    Public Function ExecuteNonQuery(sql As String, ParamArray parameters As SQLiteParameter()) As Integer
        Using cmd As New SQLiteCommand(sql, Connection)
            If parameters IsNot Nothing AndAlso parameters.Length > 0 Then
                cmd.Parameters.AddRange(parameters)
            End If
            Return cmd.ExecuteNonQuery()
        End Using
    End Function

    ''' <summary>
    ''' 执行非查询 SQL，使用简单参数值（内部使用显式 String 类型）
    ''' </summary>
    Public Function ExecuteNonQueryWithValues(sql As String, ParamArray values As Object()) As Integer
        Using cmd As New SQLiteCommand(sql, Connection)
            If values IsNot Nothing AndAlso values.Length > 0 Then
                For i = 0 To values.Length - 1
                    ' 使用显式 String 类型，避免 AddWithValue 的类型推断
                    Dim param As New SQLiteParameter($"@p{i}", DbType.String)
                    param.Value = If(values(i) Is Nothing, DirectCast(DBNull.Value, Object), values(i).ToString())
                    cmd.Parameters.Add(param)
                Next
            End If
            Return cmd.ExecuteNonQuery()
        End Using
    End Function

    ''' <summary>
    ''' 执行标量查询，使用显式类型的参数
    ''' </summary>
    Public Function ExecuteScalar(Of T)(sql As String, ParamArray parameters As SQLiteParameter()) As T
        Using cmd As New SQLiteCommand(sql, Connection)
            If parameters IsNot Nothing AndAlso parameters.Length > 0 Then
                cmd.Parameters.AddRange(parameters)
            End If
            Dim result = cmd.ExecuteScalar()
            If result Is DBNull.Value OrElse result Is Nothing Then
                Return Nothing
            End If
            Return DirectCast(result, T)
        End Using
    End Function

    ''' <summary>
    ''' 填充 DataTable，使用显式类型的参数
    ''' </summary>
    Public Function FillDataTable(sql As String, ParamArray parameters As SQLiteParameter()) As DataTable
        Dim dt As New DataTable()
        Using cmd As New SQLiteCommand(sql, Connection)
            If parameters IsNot Nothing AndAlso parameters.Length > 0 Then
                cmd.Parameters.AddRange(parameters)
            End If
            Using reader = cmd.ExecuteReader()
                dt.Load(reader)
            End Using
        End Using
        Return dt
    End Function

    ''' <summary>
    ''' 填充 DataTable，使用简单参数值（内部使用显式 String 类型）
    ''' </summary>
    Public Function FillDataTableWithValues(sql As String, ParamArray values As Object()) As DataTable
        Dim dt As New DataTable()
        Using cmd As New SQLiteCommand(sql, Connection)
            If values IsNot Nothing AndAlso values.Length > 0 Then
                For i = 0 To values.Length - 1
                    Dim param As New SQLiteParameter($"@p{i}", DbType.String)
                    param.Value = If(values(i) Is Nothing, DirectCast(DBNull.Value, Object), values(i).ToString())
                    cmd.Parameters.Add(param)
                Next
            End If
            Using reader = cmd.ExecuteReader()
                dt.Load(reader)
            End Using
        End Using
        Return dt
    End Function

    Public Function BeginTransaction() As SQLiteTransaction
        Return Connection.BeginTransaction()
    End Function

    Public Sub CreateTable(sql As String)
        ExecuteNonQuery(sql)
    End Sub

    Public Sub DropTableIfExists(tableName As String)
        ExecuteNonQuery($"DROP TABLE IF EXISTS {tableName}")
    End Sub

    Public Sub DropViewIfExists(viewName As String)
        ExecuteNonQuery($"DROP VIEW IF EXISTS {viewName}")
    End Sub

    ''' <summary>
    ''' 创建参数（显式指定类型）
    ''' </summary>
    Public Shared Function CreateParameter(name As String, dbType As DbType, value As Object) As SQLiteParameter
        Dim param As New SQLiteParameter(name, dbType)
        param.Value = If(value Is Nothing, DirectCast(DBNull.Value, Object), value)
        Return param
    End Function

    ''' <summary>
    ''' 创建 String 参数
    ''' </summary>
    Public Shared Function CreateStringParam(name As String, value As String) As SQLiteParameter
        Return CreateParameter(name, DbType.String, value)
    End Function

    ''' <summary>
    ''' 创建 Int32 参数
    ''' </summary>
    Public Shared Function CreateIntParam(name As String, value As Integer?) As SQLiteParameter
        Return CreateParameter(name, DbType.Int32, If(value.HasValue, DirectCast(value.Value, Object), Nothing))
    End Function

    ''' <summary>
    ''' 创建 Decimal 参数
    ''' </summary>
    Public Shared Function CreateDecimalParam(name As String, value As Decimal?) As SQLiteParameter
        Return CreateParameter(name, DbType.Decimal, If(value.HasValue, DirectCast(value.Value, Object), Nothing))
    End Function

    Public Sub Dispose() Implements IDisposable.Dispose
        Dispose(True)
        GC.SuppressFinalize(Me)
    End Sub

    Protected Overridable Sub Dispose(disposing As Boolean)
        If Not _isDisposed Then
            If disposing Then
                If _connection IsNot Nothing Then
                    If _connection.State = ConnectionState.Open Then
                        _connection.Close()
                    End If
                    _connection.Dispose()
                    _connection = Nothing
                End If
            End If
            _isDisposed = True
        End If
    End Sub
End Class

File: database_helper.py
"""数据库辅助类 - 封装 SQLite 数据库操作"""

import sqlite3
import threading
from typing import Any, Optional


class DatabaseHelper:
    """SQLite 数据库操作辅助类（线程安全）"""

    def __init__(self, connection_string: str):
        self._connection_string = connection_string
        self._connection: Optional[sqlite3.Connection] = None
        self._lock = threading.Lock()

    @property
    def connection(self) -> sqlite3.Connection:
        if self._connection is None:
            # check_same_thread=False 允许跨线程使用，配合 _lock 保证安全
            self._connection = sqlite3.connect(self._connection_string, check_same_thread=False)
            self._connection.execute("PRAGMA foreign_keys = ON")
        return self._connection

    def ensure_connection(self):
        _ = self.connection

    def execute_non_query(self, sql: str, params: tuple = ()) -> int:
        conn = self.connection
        with self._lock:
            # 多条语句（如 "DROP VIEW ...; CREATE VIEW ..."）需要用 executescript
            if not params and ";" in sql.strip().rstrip(";"):
                conn.executescript(sql)
                conn.commit()
                return 0
            cursor = conn.cursor()
            cursor.execute(sql, params)
            conn.commit()
            return cursor.rowcount

    def execute_scalar(self, sql: str, params: tuple = ()) -> Any:
        conn = self.connection
        with self._lock:
            cursor = conn.cursor()
            cursor.execute(sql, params)
            row = cursor.fetchone()
            if row is None:
                return None
            return row[0]

    def fill_data_table(self, sql: str, params: tuple = ()) -> tuple:
        """返回 (columns, rows) 元组，columns 为列名列表，rows 为行数据列表"""
        conn = self.connection
        with self._lock:
            cursor = conn.cursor()
            cursor.execute(sql, params)
            columns = [desc[0] for desc in cursor.description] if cursor.description else []
            rows = cursor.fetchall()
            return columns, rows

    def begin_transaction(self):
        return self.connection

    def create_table(self, sql: str):
        self.execute_non_query(sql)

    def drop_table_if_exists(self, table_name: str):
        self.execute_non_query(f"DROP TABLE IF EXISTS {table_name}")

    def drop_view_if_exists(self, view_name: str):
        self.execute_non_query(f"DROP VIEW IF EXISTS {view_name}")

    def close(self):
        with self._lock:
            if self._connection is not None:
                self._connection.close()
                self._connection = None
